### ISSUE TITLE
Copy-pasta replacement for ES Modules + better performance

### ISSUE THREAD

**Issue by samclaus** [stewartlord/identicon.js#53]
> **TL;DR**
> If you just want some copy-pasta to get 30x performance improvement and escape the Angular CommonJS warning in your project, scroll to the end of the issue.

Hi @stewartlord,

First of all, thank you for this library. We have used it for quite a while now at my place of work to distinguish users by generating default avatars from their UUIDs. It served its purpose very well for a long time, and I must prefix any constructive criticism with my long overdue appreciation!

```
 _______ _    _          _   _ _  __ _____   _____     __          _______ 
 |__   __| |  | |   /\   | \ | | |/ // ____| |  __ \   /\ \        / / ____|
    | |  | |__| |  /  \  |  \| | ' /| (___   | |  | | /  \ \  /\  / / |  __ 
    | |  |  __  | / /\ \ | . ` |  <  \___ \  | |  | |/ /\ \ \/  \/ /| | |_ |
    | |  | |  | |/ ____ \| |\  | . \ ____) | | |__| / ____ \  /\  / | |__| |
    |_|  |_|  |_/_/    \_\_| \_|_|\_\_____/  |_____/_/    \_\/  \/   \_____|
```

It's kind of sad, but like many others, the only reason I even thought to visit this project page is because I finally got sick and tired of Angular yelling at me about CommonJS dependencies. Obviously, a tiny CommonJS library like `Identicon.js` is not _really_ an issue, since there is no tree shaking to be done in the first place.

ES module usage would be nice, and so I went ahead in my usual autistic fashion and painstakingly migrated your library to TypeScript with proper class syntax and modern ES module exports...

Then I threw that code away.

While rewriting, I realized that the library does a **LOT** of jumping around between methods and even class/object boundaries and is kind of hard to reason about, which is a big red flag when it comes to performance because if a programmer has a hard time following the code, the optimizing JIT compiler probably _also_ has a hard time understanding the code and therefore cannot optimize it.

Although I have a long history of being an arrogant a\*\*hole and talking down to people, this is intended to be a constructive issue so I have whipped up some numbers for you and also some copy-pasta code that people could paste into their own projects if you would be so kind as to pin this issue (I am too busy/lazy to go make a proper NPM package right now, and I already spent a couple hours on this).

Now to the numbers. First, here is my testing code that I simply dumped in the middle of a REST API call in my project where I would have easy access to lots of hexadecimal UUIDs for use as seeds. I also peppered in some array copying with intermittent strings to make sure the JavaScript optimizer doesn't try anything fancy (i.e., remove any code) that would mess up the performance test.

```TypeScript
let seeds: string[] = [];

// allMembers is a list of membership objects from a REST API in my project, but you could substitute
// some random seed generation code and make a big array of your own seeds if you don't trust me
// and want to test yourself
for (const { id, userID, projectID } of allMembers) {
    seeds.push(id, userID, projectID);
}

// Make sure to confuse the JIT lol (the strcmp at the end is a function for sorting alphabetically
seeds = [...seeds, "0adf9940-b117-4e6c-82ad-4a345024621d", ...seeds, ...seeds, "0adf9940-b117-4e6c-82ad-4a345024621d", ...seeds].sort(strcmp);

const identicons: string[] = [];
const start = performance.now();

for (const seed of seeds) {
    identicons.push(

        // My copy-pasta replacement for your library
        genIdenticonSVG(seed, {
            margin: 0.2,
            size: 64,
            background: [255, 255, 255, 255],
        }),

        // Your library, without any tampering or funny business
        // (new Identicon(seed, {
        //     margin: 0.2,
        //     size: 64,
        //     background: [255, 255, 255, 255],
        // })).toString(),

    );
}

console.log(identicons);
console.log(`Took ${Math.floor(performance.now() - start)}ms to produce ${identicons.length} identicons!`);
```

Okay, so now that we have the testing code out of the way, let's take a look at the numbers!

**With my `genIdenticonSVG` function**

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/13316426/160312908-7c3f0286-1e0e-4c00-a536-abf5cefda83a.png">

**With the `Identicon.js` library**

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/13316426/160312947-04184d11-434c-43a2-8a82-3907e231419c.png">

I spam refreshed the memberships table in my app to re-run the API call and thus the test. With my function, the test generally took between 22 and 35 milliseconds. **It never reached 40 ms.** With the `Identicon.js` library, **it always took at least 1300 ms.** That means over a **30x performance improvement**, at least on my computer.

So how did I do it? By simplifying the code in an OCD, straightforward manner. Any time I call the `Identicon` constructor, it is never to do anything other than immediately call `toString()`--there is no reason to have an `Identicon` class in the first place. Browsers put in a great deal of work to optimize object property lookups (even on class-based objects) because technically every JS object must behave like a hashmap and can change at any time. When you put everything into a single function scope with `const` number/string variables and whatnot, and put all the loops and code _right there_ next to each other, it becomes

1. much easier to read
2. much easier for the browser to optimize (variable declarations can't change and it can more easily inline code inside the loops)

I also store rectangles as arrays/tuples (`[x, y, width, height, color]`) because, once again, object properties make JS engines suffer while array indexing is much easier for them to make guarantees about and thus optimize.

Now, _there are a couple caveats_ to my code:

1. I don't support PNG
    > I would recommend a simple function like I have to generate the rectangles, and then have separate SVG and PNG functions which both internally call the function to get the rectangles and then do their thing with them because mixing in the `{ index(); getDump(); toString() }` interface into the business code is part of what hurts the performance so much in the first place because it introduces a bunch of dynamic code into an otherwise low-level function
2. I use `btoa` without a polyfill because I have the privilege of telling users not to use Internet Explorer and I only use the code in a frontend application so no need for Node support
3. I do not have the same interface as your code, and I do not have the energy to make a proper PR to bring your library and the `PNGlib` code (which suffers similar performance problems) up-to-date

# Copy-pasta re-implementation of the library (SVG only)

At long last, here is the code. It could still use some further cleanup and perf improvements, but I figure it is good enough for now. If I publish it as a library on NPM in the future, I will probably include PNG support and I will make sure to credit you and the people credited in your code (good on you for those big comments pointing to other programmers!) so that I am contributing to open source rather than stealing.

People can easily make this faster by simply removing the options and hard-coding the values they want to use for margin, saturation, etc. If you do need to use different options with this copy-pasta garbage, you could at least generate a fill-out `options` object _once_ and pass that so you don't need to run the default (`??`) value code every time you generate an identicon with the exact same options.

```TypeScript
// adapted from: https://gist.github.com/aemkei/1325937
function hsl2rgb(h: number, s: number, b: number): [number, number, number] {
    h *= 6;

    const lookup: readonly number[] = [
        b += s *= b < .5 ? b : 1 - b,
        b - h % 1 * s * 2,
        b -= s *= 2,
        b,
        b + h % 1 * s,
        b + s
    ];

    return[
        lookup[ ~~h    % 6 ] * 255, // red
        lookup[ (h|16) % 6 ] * 255, // green
        lookup[ (h|8)  % 6 ] * 255  // blue
    ];
}

/**
 * Create a CSS color string, such as "rgba(255, 122, 37, 0.6)" from number inputs. Red,
 * green, and blue values will be floored to integers. The alpha value will be normalized
 * automatically so you can pass a decimal in the range [0, 1] OR you can pass an integer
 * in the range [0, 255].
 */
function cssColor(r: number, g: number, b: number, a: number): string {
    const int = Math.floor;
    return `rgba(${int(r)},${int(g)},${int(b)},${(a >= 0) && (a <= 255) ? a/255 : 1})`;
}

export interface Options {
    /**
     * The background color, as an RGBA tuple. The red, green, and blue values
     * should be integers in the range [0, 255] while the alpha (last number)
     * can also be an integer in that range or can be normalized to the [0, 1]
     * range (in which case it would be a decimal).
     */
    background?: [number, number, number, number];
    /**
     * Decimal margin to determine spacing between rectangles. Defaults to 0.08.
     */
    margin?: number;
    /**
     * Width and height, in pixels, of the generated SVG viewport. Defaults to 64.
     */
    size?: number;
    /**
     * Do you like color? I like color. Defaults to 0.7.
     */
    saturation?: number;
    /**
     * How bright should the computed foreground color be? Defaults to 0.5.
     */
    brightness?: number;
}

/**
 * This code was produced by simplifying/optimizing the code from Identicon.js 2.3.3
 * http://github.com/stewartlord/identicon.js
 *
 * Copyright 2018, Stewart Lord <https://github.com/stewartlord>
 * Released under the BSD license
 * http://www.opensource.org/licenses/bsd-license.php
 * 
 * Also credit to me (Sam Claus <https://github.com/samclaus>) for the rewrite/optimization.
 */
export function genIdenticonSVG(hash: string, options: Partial<Options> = {}): string {
    if (hash.length < 15) {
        throw new Error("hash for identicon must be at least 15 characters");
    }

    const bg = cssColor(...(options.background ?? [240, 240, 240, 255]));
    const size = options.size ?? 64;
    const baseMargin = Math.floor(size * (options.margin ?? 0.08));
    const saturation = options.saturation ?? 0.7;
    const brightness = options.brightness ?? 0.5;

    // foreground defaults to last 7 chars as hue at 70% saturation, 50% brightness
    const hue = parseInt(hash.substr(-7), 16) / 0xfffffff;
    const fg = cssColor(...hsl2rgb(hue, saturation, brightness), 255);
    const cell = Math.floor((size - (baseMargin * 2)) / 5);
    const margin = Math.floor((size - cell * 5) / 2);

    // X, Y, width, height, color for each rectangle
    const rectangles: [number, number, number, number, string][] = [];

    // the first 15 characters of the hash control the pixels (even/odd)
    // they are drawn down the middle first, then mirrored outwards
    for (let i = 0; i < 15; i++) {
        const color = parseInt(hash.charAt(i), 16) % 2 ? bg : fg;

        if (i < 5) {
            rectangles.push([2 * cell + margin, i * cell + margin, cell, cell, color]);
        } else if (i < 10) {
            rectangles.push([1 * cell + margin, (i - 5) * cell + margin, cell, cell, color]);
            rectangles.push([3 * cell + margin, (i - 5) * cell + margin, cell, cell, color]);
        } else if (i < 15) {
            rectangles.push([0 * cell + margin, (i - 10) * cell + margin, cell, cell, color]);
            rectangles.push([4 * cell + margin, (i - 10) * cell + margin, cell, cell, color]);
        }
    }

    const stroke = size * 0.005;

    let xml = `<svg xmlns='http://www.w3.org/2000/svg' width='${size}' height='${size}' style='background-color:${bg};'><g style='fill:${fg}; stroke:${fg}; stroke-width:${stroke};'>`;

    for (const [x, y, w, h, color] of rectangles) {
        if (color !== bg) {
            xml += `<rect x='${x}' y='${y}' width='${w}' height='${h}'/>`;
        }
    }

    xml += "</g></svg>"

    return btoa(xml);
}
```

**Comment by stewartlord** [comment]
Wow! Thanks for taking the time to write that up. I'm glad you've been able to make use of this library.

I think the bulk of the difference between the two versions is down to PNG vs. SVG creation and perhaps the additional memory consumed (roughly 7x larger for PNG). If I set this old library to use `format: 'svg'` and perform 10,000 iterations I get a **26x** improvement (`3516ms` vs `135ms`). By moving the object instantiation out of the loop, the earlier 10k test drops to `122ms` or **29x** faster which is eerily close to your function-based implementation.

In other words, I'd be cautious about attributing the speed-up to the browser's ability to optimize code. The PNG version does much more work and object instantiation appears to account for the remaining difference. I didn't benchmark your code, but I trust that it is even faster!

Please setup a repo and publish your implementation. I'd be happy to put a link to it here. This library is embarrassingly out of date!

**Comment by samclaus** [comment]
🤦‍♂️ Doh! Now I feel really dumb. I forgot to add back `format: "svg"` when I threw away my _slightly modified_ version of your library. When I test using `Identicon.js` with SVGs like I intended to, I get near-identical performance results! I severely underestimated just _how_ smart the JS optimizer is..
